edb/pgsql/resolver/resolver.py

```diff
@@ -0,0 +1,645 @@
+
+        if cols_names and len(cols_query) != len(cols_names):
+            raise errors.QueryError(
+                f'CTE alias for `{cte.name}` contains {len(cols_names)} '
+                f'columns, but the query resolves to `{len(cols_query)}` '
+                f'columns'
+            )
+        for index, col in enumerate(cols_query):
+            res_col = context.Column()
+            if cols_names:
+                res_col.name = cols_names[index]
+            else:
+                res_col.name = col.name
+            res_col.reference_as = col.name
+            result.columns.append(res_col)
+
+        ctx.scope.ctes.append(result)
+
+    return pgast.CommonTableExpr(
+        name=cte.name,
+        aliascolnames=None,
+        query=query,
+        recursive=cte.recursive,
+        materialized=cte.materialized,
+    )
+
+
+@dispatch._resolve.register
+def resolve_BaseRangeVar(
+    range_var: pgast.BaseRangeVar, *, ctx: Context
+) -> pgast.BaseRangeVar:
+    # store public alias
+    if range_var.alias:
+        ctx.scope.rel.alias = range_var.alias.aliasname
+
+    # generate internal alias
+    aliasname = ctx.names.generate_relation()
+    ctx.scope.rel.reference_as = aliasname
+    alias = pgast.Alias(aliasname=aliasname)
+    return dispatch.resolve_range_var(range_var, alias, ctx=ctx)
+
+
+@dispatch.resolve_range_var.register
+def resolve_RangeSubselect(
+    range_var: pgast.RangeSubselect,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.RangeSubselect:
+    with ctx.empty() as subctx:
+        subquery = dispatch.resolve(range_var.subquery, ctx=subctx)
+
+        result = context.Table()
+        result.name = range_var.alias.aliasname
+        result.reference_as = alias.aliasname
+        result.columns = [
+            context.Column(name=col.name, reference_as=col.name)
+            for col in subctx.scope.rel.columns
+        ]
+
+        ctx.scope.rel = result
+
+    return pgast.RangeSubselect(
+        subquery=subquery,
+        alias=alias,
+    )
+
+
+@dispatch.resolve_range_var.register
+def resolve_RelRangeVar(
+    range_var: pgast.RelRangeVar,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.RelRangeVar:
+    ctx.include_inherited = range_var.include_inherited
+
+    return pgast.RelRangeVar(
+        relation=dispatch.resolve(range_var.relation, ctx=ctx),
+        alias=alias,
+    )
+
+
+@dispatch.resolve_range_var.register
+def resolve_RangeFunction(
+    range_var: pgast.RangeFunction,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.RangeFunction:
+    raise errors.UnsupportedFeatureError('range functions are not supported')
+
+
+@dispatch.resolve_range_var.register
+def resolve_JoinExpr(
+    range_var: pgast.JoinExpr,
+    alias: pgast.Alias,
+    *,
+    ctx: Context,
+) -> pgast.JoinExpr:
+
+    with ctx.empty() as subctx:
+        larg = dispatch.resolve(range_var.larg, ctx=subctx)
+        l_rel = subctx.scope.rel
+
+    with ctx.empty() as subctx:
```

Comment:
I think that with `LATERAL` joins the join arguments can reference earlier relations, so in the case we don't want an empty relation